Let TDuration::isValid() validate durations

ajyoon opened the pull request:
`Score::cmdDoubleDuration()` fails to double any wider than a Whole Note. Similarly, `Score::cmdHalfDuration()` fails to shorten beyond 64ths. This seems to be the result of a redundant hard-coding of validating durations, possibly from an earlier version before such durations were supported. I suggest we let `TDuration::isValid()` handle this so that any future modifications to how durations are handled can be made without tracking down redundant validations.

One issue though is that this does expose the current lack of display support for 256ths (no flags are drawn). I thought to edit `TDuration::isValid()` to exclude 256ths, but I'm not sure if that will interfere with things other people are working on. 

PS: This is my first PR on any project, please advise me if I'm doing anything wrong! :)

Jojo-Schmitz commented:
Have you tested this? I can't get the shortcuts Q and W to than the lenght below 64th resp. above whole, with or without your change

Jojo-Schmitz commented:
AFAIR MuseScore supports 256th only for 128th triplets, not for direct input.

ajyoon commented:
My mistake - my master build also stops halving/doubling at 64th and Whole; updated original comment. But my build with this change does successfully go between 256th and Long. If the non-triplet/tuplet fraction for 256ths is invalid, shouldn't `isValid()` be responsible for that anyway? I can take a look later to see about it.

ajyoon commented:
I noticed in tuples it was still possible to reach a 256th, which displayed incorrectly. They should be disallowed across the board now.

(sorry for the small git mistake panic, I'll look into a way to clean that up)

Jojo-Schmitz commented on mscore/scoreview.cpp:
```diff
@@ -4821,10 +4821,11 @@ void ScoreView::setCursorVisible(bool v)
+      if (!d.isValid() && cr->durationType() != TDuration(TDuration::DurationType::V_MEASURE)) {
+           mscore->noteTooShortForTupletDialog();
+           return;
+           }
```
whitespace damage, keep the indet of 6 spaces

Jojo-Schmitz commented:
Well, with those changes I can still create a triplets containing 3 256th, aand even regular ones, and still can't use Q/W to half/double Durations below/above 64th/whole.

Jojo-Schmitz commented:
And they look OK, as Long as they are beamed, but have no flag if not

Jojo-Schmitz commented:
Ah, I see, the glyphs for flag256thUp and flag256thDown are missing in Emmentaler

ajyoon commented:
Ah! I didn't think to check if 256th's worked with beams, only was checking with stem. Should we be allowing 256ths in the first place if they don't display correctly without beams at the moment?

But anyway I don't understand, it seems like you're seeing something totally different than I am on my branch build. These were both made by repeatedly pressing Q and W. 

![wq example](https://cloud.githubusercontent.com/assets/15976214/15999528/6165dc84-3114-11e6-8ccd-960aab736ebf.png)

Attempting to change an existing 128th note into a triplet (ctrl-3) yields the expected error:
![wq example 2](https://cloud.githubusercontent.com/assets/15976214/15999529/6168cb88-3114-11e6-91ec-099d17492ea9.png)

Any idea what could be causing this discrepancy in our build behaviors?

ajyoon commented:
RE: Emmentaler missing glyph - this same error even appears in lilypond:
```
\relative c' {
    c256 d e f
}
```
![llilypond](https://cloud.githubusercontent.com/assets/15976214/15999724/91dcc82c-3115-11e6-8502-ad794a3b9f03.png)

Jojo-Schmitz commented:
No matter what I try, Q/W don't work (beyond 64th/whole), I can select a 128th chord or rest and create a triplet via Ctrl+3 and 265th flags show with Bravura (only), but there stem lenght is too short.

Jojo-Schmitz commented:
Sorry, my fault.I was building it, but not doing the install step, so executing an old version :-(

Jojo-Schmitz commented:
See https://musescore.org/en/node/43906 about 256th notes/rests and disabling 128th triplets

Jojo-Schmitz commented:
Have you signed the [CLA](https://musescore.org/cla)?
Also you'd need to squash your commits into one, maybe two (disabling 128th triplets is separate from getting Q/W to work). It'd be good to create an issue in the [issue tracker](https://musescore.org/en/node/add/project-issue/musescore) for the Q/W problem and then reference it in the commit message ("fix #nnnnnn ...")

ajyoon commented:
Yes, that's me on the CLA. Added the issue to the tracker, and rebased with commits pointing to the issues in question. How's it look?

Please note that with these commits, 256ths are not allowed under any context that I'm aware of. Meaning we can't engrave Beethoven. I'll look into getting proper support for this note value in the future. I can't believe Emmentaler doesn't support it.

ajyoon commented:
Sorry, this isn't ready to pull quite yet. Just realized (https://musescore.org/en/node/43906#comment-198036) that it's still possible to reach 256th by, for example, pressing ctrl-9 on a 32nd, causing extremely unpredictable behavior and actually trying to even repeat (R) the 256th nonuplet causes a segfault. Will investigate...

ajyoon commented:
I have a working hacky solution, but I think I have a better solution which might be controversial I'd like to run by you first: 

I think the real issue here and (I'm sure in many other places) is the entire model for `TDuration` is leaky around small note values and relies on case-by-case duration checks. I think it has to do with existing enums for unsupported note values (256, 512, 1028th). Many methods in `TDuration` legally allow for these unsupported note values to be passed without defaulting to `DurationType::V_INVALID` as they should (see `TDuration::TDuration(const Fraction& _f)`). If it's a good idea, I'd like to go and remove (or even just comment out for now) support for these enums so that any method which needs to validate a `TDuration` can just call `isValid()` to check. Then later if we decide to more robustly support 256ths it will be relatively painless to just go and modify a few lines in the enums and in `isValid()`. What do you think?

ajyoon commented:
Okay new idea - I'm reverting the work on #43906 - let's just have this PR patch up #114956 and then in a separate PR I can address #43906 by adding proper support for 256, 512, and 1024's in Emmentaler and patching up the enum check leaks since it's really a separate, bigger issue. I already have a working version with new glyphs going so I can have it finished soon enough. Sorry for all the comments!

Jojo-Schmitz commented:
the "fix #nnnnn" doesn't need to be the only thing in the commit message...

Jojo-Schmitz commented:
Looks good to me, for master and 2.0.4 ;-)

lasconic commented:
Cross referencing the CLA. https://musescore.org/en/user/647896
Thank you for your PR and welcome in the MuseScore contributors community!
